File: app/src/api/registry.ts
/**
 * Registry probe (SPEC.md §5.5) — the live half of the Pull pane.
 *
 * The bundled catalog answers "what's out there?"; this answers "does the
 * exact thing I typed exist, and how big is it?" — including models
 * published after the catalog was generated.
 *
 * `registry.ollama.ai` sends no CORS headers, so this cannot be a `fetch`
 * from the webview; the request is made in Rust (`src-tauri/src/registry.rs`)
 * and reached through the global Tauri bridge. That keeps the frontend free
 * of `@tauri-apps/*` npm dependencies, and means we hand across a model
 * *name* rather than a URL.
 *
 * Progressive enhancement, deliberately: outside the desktop shell (a plain
 * `npm run dev` browser tab, or a test run) `probeModel` resolves to
 * `unavailable` and the UI simply omits the size line. Pulling still works —
 * Ollama reports a bad name itself.
 */

export interface ProbeFound {
  kind: "found";
  /** Total download size in bytes: every layer plus the config blob. */
  totalBytes: number;
  /** Normalised `namespace/model:tag`, e.g. `library/llama3.2:latest`. */
  resolved: string;
}

/** The registry answered, and there is no such model. A real answer. */
export interface ProbeMissing {
  kind: "missing";
  resolved: string;
}

/**
 * We could not find out — offline, timeout, 5xx, or a reference we don't
 * know how to probe. Distinct from `missing` on purpose: never tell someone
 * their model doesn't exist because their wifi dropped.
 */
export interface ProbeUnknown {
  kind: "unknown";
  reason: string;
}

/** No bridge available (browser dev, tests). The UI shows nothing at all. */
export interface ProbeUnavailable {
  kind: "unavailable";
}

export type ProbeResult = ProbeFound | ProbeMissing | ProbeUnknown | ProbeUnavailable;

/** Shape of the Rust `Probe` struct, serialised camelCase. */
interface WireProbe {
  exists: boolean;
  totalBytes: number;
  resolved: string;
}

type InvokeFn = <T>(cmd: string, args?: Record<string, unknown>) => Promise<T>;

interface TauriGlobal {
  core?: { invoke?: InvokeFn };
  /** Tauri 2 betas exposed invoke at the top level; tolerate both. */
  invoke?: InvokeFn;
}

/**
 * Resolve the bridge, or null when we're not running inside the shell.
 * Looked up per call rather than cached — `withGlobalTauri` is injected
 * before our bundle runs, but a cached null would be unrecoverable if that
 * ever stopped being true.
 */
function getInvoke(): InvokeFn | null {
  if (typeof window === "undefined") return null;
  const tauri = (window as unknown as { __TAURI__?: TauriGlobal }).__TAURI__;
  return tauri?.core?.invoke ?? tauri?.invoke ?? null;
}

/** True when the probe can actually run — lets the UI skip the round trip. */
export function isProbeAvailable(): boolean {
  return getInvoke() !== null;
}

/**
 * Ask the registry about `[namespace/]model[:tag]`.
 *
 * Never rejects: every failure is a `ProbeResult` variant, because this is
 * a typing-time affordance and must not be able to break the input bar.
 */
export async function probeModel(reference: string): Promise<ProbeResult> {
  const invoke = getInvoke();
  if (invoke === null) return { kind: "unavailable" };
  if (reference.trim() === "") return { kind: "unknown", reason: "empty model name" };

  try {
    const probe = await invoke<WireProbe>("probe_model", { reference });
    return probe.exists
      ? { kind: "found", totalBytes: probe.totalBytes, resolved: probe.resolved }
      : { kind: "missing", resolved: probe.resolved };
  } catch (err) {
    // Rust returns Err(String); Tauri surfaces it as the rejection value.
    return { kind: "unknown", reason: typeof err === "string" ? err : String(err) };
  }
}

File: src-tauri/src/registry.rs
//! Registry probe for the Pull pane (SPEC.md §5.5).
//!
//! Ollama has no search API, so Remuda ships a generated catalog for
//! browsing. This covers the other half: confirming that a name the user
//! typed *actually exists* — including models published after our catalog
//! was generated — and reporting its real download size before they commit.
//!
//! This lives in Rust for two reasons. `registry.ollama.ai` sends no
//! `Access-Control-Allow-Origin`, so a webview `fetch` can't read the
//! response; and building the URL here means the frontend passes a model
//! *name*, never a URL, so this can't be turned into a general-purpose
//! request proxy.
//!
//! This is the only outbound (non-loopback) request Remuda makes.

use serde::Serialize;

const REGISTRY: &str = "https://registry.ollama.ai/v2";
const DEFAULT_NAMESPACE: &str = "library";
const DEFAULT_TAG: &str = "latest";
/// Generous next to real names (the longest in the library is ~40 chars),
/// but bounded so a pathological input can't be reflected into a URL.
const MAX_PART: usize = 128;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Probe {
    /// The manifest resolved (HTTP 200).
    pub exists: bool,
    /// Sum of every layer plus the config blob — what the pull will transfer.
    /// 0 when `exists` is false.
    pub total_bytes: u64,
    /// Normalised `namespace/model:tag` we actually asked about, so the UI can
    /// show what it resolved (e.g. `llama3.2` → `library/llama3.2:latest`).
    pub resolved: String,
}

/// One path segment of a model reference.
///
/// Deliberately stricter than the registry itself: alphanumerics plus
/// `. _ -`, non-empty, bounded, and never a relative-path token. Rejecting
/// `/`, `?`, `#`, `%` and `..` here is what keeps a crafted name from
/// escaping the path we intend or smuggling in a query string.
fn valid_part(part: &str) -> bool {
    !part.is_empty()
        && part.len() <= MAX_PART
        && part != "."
        && part != ".."
        && part
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-')
}

/// Split `[namespace/]model[:tag]` into the pieces the manifest URL needs.
fn parse_reference(reference: &str) -> Result<(String, String, String), String> {
    let reference = reference.trim();
    if reference.is_empty() {
        return Err("empty model name".into());
    }
    // A full URL is legal input to `ollama pull`, but we only know how to
    // probe the default registry. Say so rather than guessing — the pull
    // itself is unaffected.
    if reference.contains("://") {
        return Err("only registry.ollama.ai references can be probed".into());
    }

    let (path, tag) = match reference.rsplit_once(':') {
        Some((p, t)) => (p, t),
        None => (reference, DEFAULT_TAG),
    };
    let (namespace, model) = match path.split_once('/') {
        Some((ns, m)) => (ns, m),
        None => (DEFAULT_NAMESPACE, path),
    };

    // `valid_part` rejects `/`, so `a/b/c` fails here rather than silently
    // probing the wrong path.
    for part in [namespace, model, tag] {
        if !valid_part(part) {
            return Err(format!("invalid model reference: {reference}"));
        }
    }

    Ok((namespace.to_string(), model.to_string(), tag.to_string()))
}

#[derive(serde::Deserialize)]
struct Manifest {
    #[serde(default)]
    config: Option<Blob>,
    #[serde(default)]
    layers: Vec<Blob>,
}

#[derive(serde::Deserialize)]
struct Blob {
    #[serde(default)]
    size: u64,
}

/// Probe the registry for `[namespace/]model[:tag]`.
///
/// `Ok(exists: false)` is a real answer (the name isn't published), distinct
/// from `Err`, which means we couldn't find out — offline, DNS failure, a 5xx.
/// The UI must not present the latter as "this model doesn't exist".
#[tauri::command]
pub async fn probe_model(reference: String) -> Result<Probe, String> {
    let (namespace, model, tag) = parse_reference(&reference)?;
    let resolved = format!("{namespace}/{model}:{tag}");
    let url = format!("{REGISTRY}/{namespace}/{model}/manifests/{tag}");

    let client = reqwest::Client::builder()
        .user_agent(concat!("Remuda/", env!("CARGO_PKG_VERSION")))
        .timeout(std::time::Duration::from_secs(10))
        // The host is fixed and the path is built from a validated name; not
        // following redirects keeps it that way, rather than letting a 3xx
        // hand this request to somewhere we never intended to talk to.
        .redirect(reqwest::redirect::Policy::none())
        .build()
        .map_err(|e| e.to_string())?;

    let res = client.get(&url).send().await.map_err(|e| e.to_string())?;

    if res.status() == reqwest::StatusCode::NOT_FOUND {
        return Ok(Probe { exists: false, total_bytes: 0, resolved });
    }
    if !res.status().is_success() {
        return Err(format!("registry returned {}", res.status()));
    }

    let manifest: Manifest = res.json().await.map_err(|e| e.to_string())?;
    let total_bytes = manifest.layers.iter().map(|l| l.size).sum::<u64>()
        + manifest.config.map_or(0, |c| c.size);

    Ok(Probe { exists: true, total_bytes, resolved })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bare_name_defaults_namespace_and_tag() {
        let (ns, m, t) = parse_reference("llama3.2").unwrap();
        assert_eq!((ns.as_str(), m.as_str(), t.as_str()), ("library", "llama3.2", "latest"));
    }

    #[test]
    fn explicit_tag_and_namespace() {
        let (ns, m, t) = parse_reference("some-user/my-model:q4_K_M").unwrap();
        assert_eq!((ns.as_str(), m.as_str(), t.as_str()), ("some-user", "my-model", "q4_K_M"));
    }

    #[test]
    fn rejects_path_traversal_and_url_smuggling() {
        for bad in [
            "../../etc/passwd",
            "library/../../evil",
            "..",
            "a/b/c",
            "model?query=1",
            "model#frag",
            "model%2f",
            "http://evil.test/x",
            "",
            "   ",
        ] {
            assert!(parse_reference(bad).is_err(), "should reject {bad:?}");
        }
    }

    #[test]
    fn rejects_overlong_part() {
        assert!(parse_reference(&"a".repeat(MAX_PART + 1)).is_err());
    }

    /// Hits the real registry. Ignored by default so the suite stays offline;
    /// run with `cargo test -- --ignored --nocapture` when changing the
    /// manifest handling.
    #[test]
    #[ignore]
    fn live_probe_round_trip() {
        let found = tauri::async_runtime::block_on(probe_model("llama3.2".into())).unwrap();
        assert!(found.exists);
        assert_eq!(found.resolved, "library/llama3.2:latest");
        // ~2 GB; assert a broad band so a re-tag upstream doesn't fail this.
        assert!(
            found.total_bytes > 1_000_000_000 && found.total_bytes < 5_000_000_000,
            "unexpected size {}",
            found.total_bytes
        );

        let missing =
            tauri::async_runtime::block_on(probe_model("definitely-not-a-model-xyz".into()))
                .unwrap();
        assert!(!missing.exists);
        assert_eq!(missing.total_bytes, 0);

        println!("llama3.2 => {} bytes", found.total_bytes);
    }
}
